source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,583 @@
+                    const Tensor<2, dim, double> projection = unit_symmetric_tensor<dim>() -
+                                                              outer_product(fs_fe_face_values.normal_vector(point), fs_fe_face_values.normal_vector(point));
+
+                    // Loop over the shape functions
+                    for (unsigned int i=0; i<dofs_per_cell; ++i)
+                      {
+                        // Make sure we only assemble for the y-component
+                        // TODO this is only correct for box geometries
+                        if (mesh_deformation_dof_handler.get_fe().system_to_component_index(i).first != dim-1)
+                          continue;
+
+                        // Assemble the RHS
+                        // RHS = M*H_old
+                        cell_vector(i) += displacement *
+                                          fs_fe_face_values.shape_value (i, point) *
+                                          fs_fe_face_values.JxW(point);
+
+
+                        for (unsigned int j=0; j<dofs_per_cell; ++j)
+                          {
+                            // Make sure we only assemble for the y-component
+                            // TODO this is only correct for box geometries
+                            if (mesh_deformation_dof_handler.get_fe().system_to_component_index(j).first != dim-1)
+                              continue;
+                            // Assemble the matrix, for backward first order time discretization:
+                            // Matrix := (M+dt*K) = (M+dt*B^T*kappa*B)
+                            cell_matrix(i,j) +=
+                              (
+                                this->get_timestep() * diffusivity *
+                                projection * fs_fe_face_values.shape_grad(i, point) * projection * fs_fe_face_values.shape_grad(j,point) +
+                                fs_fe_face_values.shape_value (i, point) * fs_fe_face_values.shape_value (j, point)
+                              )
+                              * fs_fe_face_values.JxW(point);
+                          }
+                      }
+
+                  }
+
+                mass_matrix_constraints.distribute_local_to_global (cell_matrix, cell_vector,
+                                                                    cell_dof_indices, mass_matrix, system_rhs, false);
+              }
+
+      system_rhs.compress (VectorOperation::add);
+      mass_matrix.compress(VectorOperation::add);
+
+      // Jacobi seems to be fine here.  Other preconditioners (ILU, IC) run into trouble
+      // because the matrix is mostly empty, since we don't touch internal vertices.
+      LinearAlgebra::PreconditionJacobi preconditioner_mass;
+      LinearAlgebra::PreconditionJacobi::AdditionalData preconditioner_control(1,1e-16,1);
+      preconditioner_mass.initialize(mass_matrix, preconditioner_control);
```

Comment:
Yes, I haven't tried any other preconditioners, because of the comment (taken from the free surface implementation) that they won't like the zero matrix entries. But I could look into that.